### PR TITLE
Fix async inserts with empty data

### LINKED ISSUE #48545: Unexpected exception while inserting empty data with async_insert=1
**Unexpected behaviour**
Insertion of the empty data set (without records) causes an unexpected and unclear error when async_insert is enabled, error: `The associated promise has been destructed prior to the associated state becoming ready`.
See the stack trace below. The same error will be visible on the client side.

**How to reproduce**
* ClickHouse starting from 22.12 (22.12.6.22)
* HTTP interface (curl, clickhouse/client@0.0.11)
* async_insert should be enabled
* Table for demonstration:
```
create table default.async_insert_issue (
    key String
  , data UInt32
)
engine = MergeTree() order by (key);
```
* Curl command for demonstration:
```
echo '' | curl 'http://localhost:8123?query=insert%20into%20default.async_insert_issue%20settings%20async_insert%3D1%20format%20JSONEachRow' --data-binary @-
```

**Expected behavior**
Insertion requests without data should be silently ignored, or the user should be warned that no data was provided.
Should be no unexpected errors.

**Error message and/or stacktrace**
```
2023.04.07 10:34:54.670440 [ 285 ] {3f869f30-6ef5-4430-8eff-4b692f6c97e8} <Error> DynamicQueryHandler: std::exception. Code: 1001, type: std::__1::future_error, e.what() = The associated promise has been destructed prior to the associated state becoming ready., Stack trace (when copying this message, always include the lines below):

0. std::__1::promise<void>::~promise() @ 0x13b0a7ac in /usr/bin/clickhouse
1. ? @ 0xfa01778 in /usr/bin/clickhouse
2. ? @ 0xfa00e84 in /usr/bin/clickhouse
3. DB::AsynchronousInsertQueue::processData(DB::AsynchronousInsertQueue::InsertQuery, std::__1::unique_ptr<DB::AsynchronousInsertQueue::InsertData, std::__1::default_delete<DB::AsynchronousInsertQueue::InsertData>>, std::__1::shared_ptr<DB::Context const>) @ 0xf9feff4 in /usr/bin/clickhouse
4. ? @ 0xfa01098 in /usr/bin/clickhouse
5. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0xbabb434 in /usr/bin/clickhouse
6. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) @ 0xbabdafc in /usr/bin/clickhouse
7. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xbab7bf0 in /usr/bin/clickhouse
8. ? @ 0xbabcae8 in /usr/bin/clickhouse
9. start_thread @ 0x7624 in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
10. ? @ 0xd149c in /usr/lib/aarch64-linux-gnu/libc-2.31.so
 (version 23.2.5.46 (official build))
```

PS: Actually not a big deal, but logs are flooded with those errors.

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed asynchronous inserts in case when empty data is passed.

Fixes #48545.

### CODE REVIEW

File: src/Interpreters/AsynchronousInsertQueue.cpp [hunk 1]
```diff
@@ -473,8 +480,27 @@ try
...
+        {
+            if (!entry->isFinished())
+                entry->finish();
+        }
+
+        if (!log_elements.empty())
+        {
+            auto flush_time = has_data ? std::chrono::system_clock::now() : TimePoint{};
+            auto query_id = has_data ? insert_query_id : "";
+            appendElementsToLogSafe(*insert_log, std::move(log_elements), flush_time, query_id, "");
```
Comment (yakov-olkhovskiy) [inline src/Interpreters/AsynchronousInsertQueue.cpp:495]: why flush_time and query_id should be empty if no data?

Comment (CurtizJ) [reply src/Interpreters/AsynchronousInsertQueue.cpp:495]: Because the query is not executed in case of no data.

Comment (yakov-olkhovskiy) [reply src/Interpreters/AsynchronousInsertQueue.cpp:495]: Then why log it at all? And if log it - why mess with flush time? We are actually flushing it. And the same for query id - I think it's here to identify which insert query is responsible for this flushing...
I think if we are actually flushing it these info should be preserved with no regards if data is present or not...

Comment (CurtizJ) [reply src/Interpreters/AsynchronousInsertQueue.cpp:495]: I agree, let's add this info too.

**Review by yakov-olkhovskiy — APPROVED** [review]

**Pull request merged**